lib/api/objectPut.js

```diff
@@ -46,6 +46,10 @@ function _storeIt(bucketName, objectKey, objMD, authInfo, canonicalID,
         dataToDelete = Array.isArray(objMD.location) ?
             objMD.location : [objMD.location];
     }
+    // not deleting old data in the case of versioning
+    if (objMD && objMD['x-amz-version-id'] !== 'null') {
+        dataToDelete = undefined;
+    }
```

Comment:
To be backwards compatible, shouldn't you do:
`if(objMD && objMD['x-amz-version-id'] && objMD['x-amz-version-id'] !== 'null')`

This assumes that you will be setting the objMD['x-amz-version-id'] to null for all non-versioned objects.